diff -up Python-2.6.5/Lib/urllib2.py.urllib2-ftp-no-proxy Python-2.6.5/Lib/urllib2.py
--- Python-2.6.5/Lib/urllib2.py.urllib2-ftp-no-proxy	2010-03-04 10:15:37.000000000 -0500
+++ Python-2.6.5/Lib/urllib2.py	2011-01-21 11:13:11.587525393 -0500
@@ -704,6 +704,8 @@ class ProxyHandler(BaseHandler):
         if proxy_type is None:
             proxy_type = orig_type
 
+        req.get_host()
+
         if req.host and proxy_bypass(req.host):
             return None
 
